llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -592,27 +718,311 @@ static Error applyAddends(MCCASReader &Reader,
+      uint64_t PrevOffset = CUOffset;
+      DWARFExtractor.getULEB128(&CUOffset, &Err);
+      FormSize += CUOffset - PrevOffset + 4;
+      break;
+    }
+    case dwarf::DW_FORM_string: {
+      DWARFDataExtractor DWARFExtractor(CUData, IsLittleEndian, AddressSize);
+      auto CurrOffset = CUOffset;
+      DWARFExtractor.getCStr(&CUOffset, &Err);
+      FormSize += CUOffset - CurrOffset;
+      break;
+    }
+    case dwarf::DW_FORM_indirect: {
+      DWARFDataExtractor DWARFExtractor(CUData, IsLittleEndian, AddressSize);
+      uint64_t PrevOffset = CUOffset;
+      FormVal =
+          static_cast<dwarf::Form>(DWARFExtractor.getULEB128(&CUOffset, &Err));
+      Indirect = true;
+      FormSize += CUOffset - PrevOffset;
+      break;
+    }
+    case dwarf::DW_FORM_strp:
+    case dwarf::DW_FORM_sec_offset:
+    case dwarf::DW_FORM_GNU_ref_alt:
+    case dwarf::DW_FORM_GNU_strp_alt:
+    case dwarf::DW_FORM_line_strp:
+    case dwarf::DW_FORM_strp_sup: {
+      FormSize += FP.getDwarfOffsetByteSize();
+      break;
+    }
+    case dwarf::DW_FORM_addrx3:
+    case dwarf::DW_FORM_lo_user: {
+      llvm_unreachable("usupported form");
+      break;
+    }
+    }
+  } while (Indirect && !Err);
+
+  if (Err)
+    return std::move(Err);
+
+  return FormSize;
+}
+
+static Error materializeCUDie(DWARFCompileUnit &DCU,
+                              MutableArrayRef<char> SectionContents,
+                              StringRef CUData,
+                              ArrayRef<char> FixupsDataArrayRef,
+                              uint64_t &CUOffset, uint64_t &FixupOffset,
+                              uint64_t &SectionOffset) {
+  // Copy Abbrev Tag.
+  DWARFDataExtractor DWARFExtractor(CUData, DCU.isLittleEndian(),
+                                    DCU.getAddressByteSize());
+  uint64_t PrevOffset = CUOffset;
+  uint64_t AbbrevTag = DWARFExtractor.getULEB128(&CUOffset);
+  std::memcpy(&SectionContents[SectionOffset], &AbbrevTag,
+              CUOffset - PrevOffset);
+  SectionOffset += CUOffset - PrevOffset;
+
+  // Return if the abbrev tag is 0, this indicates the end of a sequence of
+  // children DWARFDies
+  if (AbbrevTag == 0)
+    return Error::success();
+
+  // Create a dummy DWARFDie to extract the attributes.
+  DWARFDebugInfoEntry DDIE;
+  auto *AbbrevDecl =
+      DCU.getAbbreviations()->getAbbreviationDeclaration(AbbrevTag);
+  DDIE.setAbbrevDecl(AbbrevDecl);
+  DWARFDie CUDie(&DCU, &DDIE);
+
+  for (unsigned I = 0; I < AbbrevDecl->getNumAttributes(); I++) {
+    auto Form = AbbrevDecl->getFormByIndex(I);
+    auto *U = CUDie.getDwarfUnit();
+    dwarf::FormParams FP = U->getFormParams();
+    auto FormSize = getFormSize(Form, FP, CUData, CUOffset,
+                                DCU.isLittleEndian(), DCU.getAddressByteSize());
+
+    if (!FormSize)
+      consumeError(FormSize.takeError());
+
+    if (Form == llvm::dwarf::DW_FORM_strp ||
+        Form == llvm::dwarf::DW_FORM_sec_offset) {
```

Comment:
Note how we are repeating ourselves when we have this list inlined here. I left a separate comment that may help with the repetition.